From b65a17037589ebdde7708d841a1fdd86dd06ddb0 Mon Sep 17 00:00:00 2001
From: Salvo Giangreco <giangrecosalvo9@gmail.com>
Date: Mon, 6 Oct 2025 00:49:32 +0200
Subject: [PATCH] Introduce KnoxPatchHooks

UN1CA: this patch is not complete! Please check unica/mods/knoxpatch/customize.sh for more info.
---
 .../android/os/SystemProperties.smali         |  26 +-
 .../io/mesalabs/unica/KnoxPatchHooks.smali    | 269 ++++++++++++++++++
 2 files changed, 289 insertions(+), 6 deletions(-)
 create mode 100644 smali_classes6/io/mesalabs/unica/KnoxPatchHooks.smali

diff --git a/smali_classes3/android/os/SystemProperties.smali b/smali_classes3/android/os/SystemProperties.smali
index 3af84b7b..09f83663 100644
--- a/smali_classes3/android/os/SystemProperties.smali
+++ b/smali_classes3/android/os/SystemProperties.smali
@@ -373,27 +373,41 @@
 .end method
 
 .method public static whitelist get(Ljava/lang/String;)Ljava/lang/String;
-    .locals 0
+    .locals 1
     .annotation runtime Landroid/annotation/SystemApi;
     .end annotation
 
+    invoke-static {p0}, Lio/mesalabs/unica/KnoxPatchHooks;->onSystemPropertiesGet(Ljava/lang/String;)Ljava/lang/String;
+
+    move-result-object v0
+
+    if-nez v0, :cond_0
+
     invoke-static {p0}, Landroid/os/SystemProperties;->native_get(Ljava/lang/String;)Ljava/lang/String;
 
-    move-result-object p0
+    move-result-object v0
 
-    return-object p0
+    :cond_0
+    return-object v0
 .end method
 
 .method public static whitelist get(Ljava/lang/String;Ljava/lang/String;)Ljava/lang/String;
-    .locals 0
+    .locals 1
     .annotation runtime Landroid/annotation/SystemApi;
     .end annotation
 
+    invoke-static {p0}, Lio/mesalabs/unica/KnoxPatchHooks;->onSystemPropertiesGet(Ljava/lang/String;)Ljava/lang/String;
+
+    move-result-object v0
+
+    if-nez v0, :cond_0
+
     invoke-static {p0, p1}, Landroid/os/SystemProperties;->native_get(Ljava/lang/String;Ljava/lang/String;)Ljava/lang/String;
 
-    move-result-object p0
+    move-result-object v0
 
-    return-object p0
+    :cond_0
+    return-object v0
 .end method
 
 .method public static whitelist getBoolean(Ljava/lang/String;Z)Z
diff --git a/smali_classes6/io/mesalabs/unica/KnoxPatchHooks.smali b/smali_classes6/io/mesalabs/unica/KnoxPatchHooks.smali
new file mode 100644
index 00000000..8b68148a
--- /dev/null
+++ b/smali_classes6/io/mesalabs/unica/KnoxPatchHooks.smali
@@ -0,0 +1,269 @@
+.class public final Lio/mesalabs/unica/KnoxPatchHooks;
+.super Ljava/lang/Object;
+.source "KnoxPatchHooks.java"
+
+
+# static fields
+.field private static final blacklist DEBUG:Z = false
+
+.field private static final blacklist FIND_MY_MOBILE_PACKAGE_NAME:Ljava/lang/String; = "com.samsung.android.fmm"
+
+.field private static final blacklist SAMSUNG_CLOUD_PLATFORM_MANAGER_PACKAGE_NAME:Ljava/lang/String; = "com.samsung.android.scpm"
+
+.field private static final blacklist SAMSUNG_HEALTH_PACKAGE_NAME:Ljava/lang/String; = "com.sec.android.app.shealth"
+
+.field private static final blacklist SAMSUNG_TV_PLUS_PACKAGE_NAME:Ljava/lang/String; = "com.samsung.android.tvplus"
+
+.field private static final blacklist TAG:Ljava/lang/String; = "KnoxPatchHooks"
+
+.field private static volatile blacklist sPackageName:Ljava/lang/String;
+
+.field private static volatile blacklist sSpoofBuildTypeProp:Z
+
+.field private static volatile blacklist sSpoofKeystoreProp:Z
+
+.field private static volatile blacklist sSpoofKnox:Z
+
+
+# direct methods
+.method private constructor blacklist <init>()V
+    .locals 0
+
+    invoke-direct {p0}, Ljava/lang/Object;-><init>()V
+
+    return-void
+.end method
+
+.method private static blacklist dlog(Ljava/lang/String;)V
+    .locals 0
+
+    return-void
+.end method
+
+.method public static blacklist init(Landroid/content/Context;)V
+    .locals 5
+
+    invoke-virtual {p0}, Landroid/content/Context;->getPackageName()Ljava/lang/String;
+
+    move-result-object v0
+
+    invoke-static {v0}, Landroid/text/TextUtils;->isEmpty(Ljava/lang/CharSequence;)Z
+
+    move-result v1
+
+    if-eqz v1, :cond_0
+
+    return-void
+
+    :cond_0
+    sput-object v0, Lio/mesalabs/unica/KnoxPatchHooks;->sPackageName:Ljava/lang/String;
+
+    const/4 v1, 0x1
+
+    const/4 v2, 0x0
+
+    :try_start_0
+    invoke-virtual {p0}, Landroid/content/Context;->getPackageManager()Landroid/content/pm/PackageManager;
+
+    move-result-object p0
+
+    invoke-virtual {p0, v0, v2}, Landroid/content/pm/PackageManager;->getApplicationInfo(Ljava/lang/String;I)Landroid/content/pm/ApplicationInfo;
+
+    move-result-object p0
+
+    iget p0, p0, Landroid/content/pm/ApplicationInfo;->flags:I
+
+    and-int/lit16 p0, p0, 0x81
+
+    if-nez p0, :cond_1
+
+    move p0, v1
+
+    goto :goto_0
+
+    :cond_1
+    move p0, v2
+
+    :goto_0
+    sput-boolean p0, Lio/mesalabs/unica/KnoxPatchHooks;->sSpoofKnox:Z
+    :try_end_0
+    .catch Ljava/lang/Exception; {:try_start_0 .. :try_end_0} :catch_0
+
+    goto :goto_1
+
+    :catch_0
+    move-exception p0
+
+    new-instance v3, Ljava/lang/StringBuilder;
+
+    const-string v4, "sSpoofKnox = false; "
+
+    invoke-direct {v3, v4}, Ljava/lang/StringBuilder;-><init>(Ljava/lang/String;)V
+
+    invoke-virtual {p0}, Ljava/lang/Object;->getClass()Ljava/lang/Class;
+
+    move-result-object p0
+
+    invoke-virtual {p0}, Ljava/lang/Class;->getName()Ljava/lang/String;
+
+    move-result-object p0
+
+    invoke-virtual {v3, p0}, Ljava/lang/StringBuilder;->append(Ljava/lang/String;)Ljava/lang/StringBuilder;
+
+    move-result-object p0
+
+    invoke-virtual {p0}, Ljava/lang/StringBuilder;->toString()Ljava/lang/String;
+
+    move-result-object p0
+
+    invoke-static {p0}, Lio/mesalabs/unica/KnoxPatchHooks;->dlog(Ljava/lang/String;)V
+
+    sput-boolean v2, Lio/mesalabs/unica/KnoxPatchHooks;->sSpoofKnox:Z
+
+    :goto_1
+    const-string p0, "com.samsung.android.scpm"
+
+    invoke-virtual {v0, p0}, Ljava/lang/String;->equals(Ljava/lang/Object;)Z
+
+    move-result p0
+
+    sput-boolean p0, Lio/mesalabs/unica/KnoxPatchHooks;->sSpoofBuildTypeProp:Z
+
+    const-string p0, "com.samsung.android.fmm"
+
+    invoke-virtual {v0, p0}, Ljava/lang/String;->equals(Ljava/lang/Object;)Z
+
+    move-result p0
+
+    if-nez p0, :cond_3
+
+    const-string p0, "com.sec.android.app.shealth"
+
+    invoke-virtual {v0, p0}, Ljava/lang/String;->equals(Ljava/lang/Object;)Z
+
+    move-result p0
+
+    if-nez p0, :cond_3
+
+    const-string p0, "com.samsung.android.tvplus"
+
+    invoke-virtual {v0, p0}, Ljava/lang/String;->equals(Ljava/lang/Object;)Z
+
+    move-result p0
+
+    if-eqz p0, :cond_2
+
+    goto :goto_2
+
+    :cond_2
+    move v1, v2
+
+    :cond_3
+    :goto_2
+    sput-boolean v1, Lio/mesalabs/unica/KnoxPatchHooks;->sSpoofKeystoreProp:Z
+
+    return-void
+.end method
+
+.method public static blacklist onEDMGetAPILevel()I
+    .locals 1
+
+    sget-boolean v0, Lio/mesalabs/unica/KnoxPatchHooks;->sSpoofKnox:Z
+
+    if-nez v0, :cond_0
+
+    const-string v0, "39"
+
+    invoke-static {v0}, Ljava/lang/Integer;->parseInt(Ljava/lang/String;)I
+
+    move-result v0
+
+    return v0
+
+    :cond_0
+    const-string v0, "Spoofing EnterpriseDeviceManager.getAPILevel()"
+
+    invoke-static {v0}, Lio/mesalabs/unica/KnoxPatchHooks;->dlog(Ljava/lang/String;)V
+
+    const/4 v0, -0x1
+
+    return v0
+.end method
+
+.method public static blacklist onSystemPropertiesGet(Ljava/lang/String;)Ljava/lang/String;
+    .locals 3
+
+    const-string v0, "ro.build.type"
+
+    invoke-virtual {p0, v0}, Ljava/lang/String;->equals(Ljava/lang/Object;)Z
+
+    move-result v0
+
+    if-eqz v0, :cond_0
+
+    sget-boolean v0, Lio/mesalabs/unica/KnoxPatchHooks;->sSpoofBuildTypeProp:Z
+
+    if-eqz v0, :cond_1
+
+    const-string v0, "eng"
+
+    goto :goto_0
+
+    :cond_0
+    const-string v0, "ro.security.keystore.keytype"
+
+    invoke-virtual {p0, v0}, Ljava/lang/String;->equals(Ljava/lang/Object;)Z
+
+    move-result v0
+
+    if-eqz v0, :cond_1
+
+    sget-boolean v0, Lio/mesalabs/unica/KnoxPatchHooks;->sSpoofKeystoreProp:Z
+
+    if-eqz v0, :cond_1
+
+    const-string v0, ""
+
+    goto :goto_0
+
+    :cond_1
+    const/4 v0, 0x0
+
+    :goto_0
+    if-eqz v0, :cond_2
+
+    new-instance v1, Ljava/lang/StringBuilder;
+
+    const-string v2, "Spoofing \""
+
+    invoke-direct {v1, v2}, Ljava/lang/StringBuilder;-><init>(Ljava/lang/String;)V
+
+    invoke-virtual {v1, p0}, Ljava/lang/StringBuilder;->append(Ljava/lang/String;)Ljava/lang/StringBuilder;
+
+    move-result-object p0
+
+    const-string v1, "\" prop to \""
+
+    invoke-virtual {p0, v1}, Ljava/lang/StringBuilder;->append(Ljava/lang/String;)Ljava/lang/StringBuilder;
+
+    move-result-object p0
+
+    invoke-virtual {p0, v0}, Ljava/lang/StringBuilder;->append(Ljava/lang/String;)Ljava/lang/StringBuilder;
+
+    move-result-object p0
+
+    const-string v1, "\""
+
+    invoke-virtual {p0, v1}, Ljava/lang/StringBuilder;->append(Ljava/lang/String;)Ljava/lang/StringBuilder;
+
+    move-result-object p0
+
+    invoke-virtual {p0}, Ljava/lang/StringBuilder;->toString()Ljava/lang/String;
+
+    move-result-object p0
+
+    invoke-static {p0}, Lio/mesalabs/unica/KnoxPatchHooks;->dlog(Ljava/lang/String;)V
+
+    :cond_2
+    return-object v0
+.end method
-- 
2.51.0

